File: app/services/work_orders.py
"""Work order service layer aligned with the live Supabase schema."""

from __future__ import annotations

from datetime import datetime

from sqlalchemy.orm import Session

from app.core.errors import ErrorCode, ProblemException
from app.models.work_order import WorkOrder
from app.repositories import work_orders as wo_repo
from app.schemas.work_orders import WorkOrderCreate, WorkOrderUpdate


# PUBLIC_INTERFACE
def create_work_order(db: Session, payload: WorkOrderCreate) -> WorkOrder:
    """Create a work order from live request fields.

    Args:
        db: Active database session.
        payload: Validated work-order creation payload.

    Returns:
        WorkOrder: The persisted work order.
    """
    work_order = WorkOrder(
        equipment_id=payload.equipment_id,
        title=payload.title,
        description=payload.description,
        priority=payload.priority.value,
        status="OPEN",
        assigned_to=None,
        closed_by=None,
    )
    wo_repo.add(db, work_order)
    db.commit()
    db.refresh(work_order)
    return wo_repo.get_by_id(db, work_order.id) or work_order


# PUBLIC_INTERFACE
def get_work_order(db: Session, work_order_id: str) -> WorkOrder:
    """Return a work order by id or raise a not-found problem."""
    work_order = wo_repo.get_by_id(db, work_order_id)
    if work_order is None:
        raise ProblemException(
            status=404,
            code=ErrorCode.WORK_ORDER_NOT_FOUND,
            detail=f"Work order '{work_order_id}' does not exist.",
        )
    return work_order


# PUBLIC_INTERFACE
def update_work_order(
    db: Session, work_order_id: str, payload: WorkOrderUpdate
) -> WorkOrder:
    """Update only live work-order columns."""
    work_order = get_work_order(db, work_order_id)
    data = payload.model_dump(exclude_unset=True)

    if "title" in data and data["title"] is not None:
        work_order.title = data["title"]
    if "description" in data and data["description"] is not None:
        work_order.description = data["description"]
    if "priority" in data and payload.priority is not None:
        work_order.priority = payload.priority.value
    if "status" in data and payload.status is not None:
        work_order.status = payload.status.value
    if "assigned_to" in data:
        work_order.assigned_to = data["assigned_to"]
    if "closed_by" in data:
        work_order.closed_by = data["closed_by"]

    db.commit()
    db.refresh(work_order)
    return wo_repo.get_by_id(db, work_order.id) or work_order


# PUBLIC_INTERFACE
def list_work_orders(
    db: Session,
    *,
    page: int,
    page_size: int,
    status: str | None = None,
    priority: str | None = None,
    created_from: datetime | None = None,
    created_to: datetime | None = None,
) -> tuple[list[WorkOrder], int]:
    """List work orders with filtering and pagination."""
    return wo_repo.list_work_orders(
        db,
        page=page,
        page_size=page_size,
        status=status,
        priority=priority,
        created_from=created_from,
        created_to=created_to,
    )

File: frontend/src/types/workOrders.ts
/**
 * Types mirroring the backend Work Order contract.
 *
 * Source of truth:
 * - `Maintanance-predictor_bootcamp/app/schemas/work_orders.py`
 * - `Maintanance-predictor_bootcamp/app/schemas/common.py`
 *
 * IMPORTANT (live Supabase schema):
 * - `public.work_orders` has NO `alert_id` column (no FK link to alerts).
 * - There are no persisted part lines.
 * - There are no closure metadata fields (resolution_notes, root_cause, closed_at).
 * - There is no due_at column on the live table.
 */

// PUBLIC_INTERFACE
export type Priority = 'CRITICAL' | 'HIGH' | 'MEDIUM';
/** Work order priority values accepted/returned by the backend. */

// PUBLIC_INTERFACE
export type WorkOrderStatus = 'OPEN' | 'CLOSED';
/** Work order lifecycle status values accepted/returned by the backend. */

// PUBLIC_INTERFACE
export const PRIORITIES: Priority[] = ['CRITICAL', 'HIGH', 'MEDIUM'];
/** Ordered list of priorities for select inputs and filters. */

// PUBLIC_INTERFACE
export const WORK_ORDER_STATUSES: WorkOrderStatus[] = ['OPEN', 'CLOSED'];
/** Ordered list of statuses for select inputs and filters. */

// PUBLIC_INTERFACE
export interface WorkOrder {
  id: string;
  equipment_id?: string | null;
  equipment_name?: string | null;
  work_order_number: number;
  title: string;
  description?: string | null;
  priority: Priority;
  status: WorkOrderStatus;
  assigned_to?: string | null;
  closed_by?: string | null;
  created_at: string;
  updated_at: string;
}
/** Full work order representation (detail / create / update responses). */

// PUBLIC_INTERFACE
export interface WorkOrderSummary {
  id: string;
  equipment_id?: string | null;
  equipment_name?: string | null;
  work_order_number: number;
  title: string;
  priority: Priority;
  status: WorkOrderStatus;
  created_at: string;
}
/** Condensed work order representation used in list responses. */

// PUBLIC_INTERFACE
export interface WorkOrderListResponse {
  items: WorkOrderSummary[];
  total: number;
  page: number;
  page_size: number;
}
/** Paginated list payload returned by `GET /v1/work-orders`. */

// PUBLIC_INTERFACE
export interface WorkOrderListParams {
  page?: number;
  page_size?: number;
  status?: WorkOrderStatus | '';
  priority?: Priority | '';
  created_from?: string;
  created_to?: string;
}
/** Query parameters supported by the work orders list endpoint. */

// PUBLIC_INTERFACE
export interface WorkOrderCreatePayload {
  /** Required title; backend persists to public.work_orders. */
  title: string;
  /** Optional equipment id; may be ignored by the create-from-alert route. */
  equipment_id?: string | null;
  /** Optional description; the create-from-alert route may append alert context. */
  description?: string | null;
  priority: Priority;
}
/** Request body for `POST /v1/alerts/{alert_id}/work-orders`. */

// PUBLIC_INTERFACE
export interface WorkOrderUpdatePayload {
  title?: string;
  description?: string | null;
  priority?: Priority;
  status?: WorkOrderStatus;
  assigned_to?: string | null;
  closed_by?: string | null;
}
/** Request body for `PUT /v1/work-orders/{work_order_id}` (at least one field). */
